# Bài Thực Hành R - Chương 2
### Các thành viên nhóm:
* Phạm Xuân Sang - 20195991
* Vũ Thành Huy - 20195968
* Hà Huyền Vi - 20196007

* **Nhập dữ liệu:**

In [28]:
library(BSDA)
abalone <- read.csv("abalone.csv")
head(abalone)
dim(abalone)

sex length diameter height weight.whole weight.shucked weight.viscera
1 M   0.455  0.365    0.095  0.5140       0.2245         0.1010        
2 M   0.350  0.265    0.090  0.2255       0.0995         0.0485        
3 F   0.530  0.420    0.135  0.6770       0.2565         0.1415        
4 M   0.440  0.365    0.125  0.5160       0.2155         0.1140        
5 I   0.330  0.255    0.080  0.2050       0.0895         0.0395        
6 I   0.425  0.300    0.095  0.3515       0.1410         0.0775        
  weight.shell rings
1 0.150        15   
2 0.070         7   
3 0.210         9   
4 0.155        10   
5 0.055         7   
6 0.120         8

[1] 4177    9

## I. Tìm khoảng tin cậy:
**1. Khoảng tin cậy cho giá trị trung bình:**


**a. Phương sai chưa biết:**
    
- Áp dụng tìm khoảng tin cậy cho cột **`length`** với độ tin cậy $\gamma = 0.99$.

>- **Cách 1:** Sử dụng công thức: 

$$P\left(\overline{x} - t^{(n-1)}_{1-\alpha_2}\dfrac{s}{\sqrt{n}} \mu \leq \overline{x} + t^{(n-1)}_{1-\alpha_1}\dfrac{s}{\sqrt{n}}\right) = 1 - \alpha$$

Tính phương sai mẫu hiệu chỉnh và Sai số tiêu chuẩn:

In [4]:
options(digits = 5)
n = length(abalone$length)
sprintf("The number of observations: %d", n)
s = sd(abalone$length)
sprintf("Sample Standard Deviation: %.5f", s)
SE = s/sqrt(n)
sprintf("Standard Error Estimate: %.5f", SE)

[1] "The number of observations: 4177"

[1] "Sample Standard Deviation: 0.12009"

[1] "Standard Error Estimate: 0.00186"

* **Ta thu được**: 
 * Phương sai mẫu hiệu chỉnh: 0.12009
 * Sai số tiêu chuẩn: 0.00186

>> Khoảng tin cậy đối xứng:

In [5]:
E = qt(.995, df = n-1)*SE;
xbar = mean(abalone$length)
sprintf("xbar: %.5f", xbar)
print(xbar + c(-E, E))

[1] "xbar: 0.52399"

[1] 0.51920 0.52878


**Kết luận:** Khoảng tin cậy đối xứng cho giá trị trung bình của chiều dài (cột **`length`**) các cá thể bào ngư là $(0.51920, 0.52878)$ với độ tin cậy 99%

>> Khoảng tin cậy trái:

In [6]:
print(c(-Inf, xbar + qt(.99, df = n-1)*SE))

[1]    -Inf 0.52832


**Kết luận:** Khoảng tin cậy đối trái cho giá trị trung bình của chiều dài (cột **`length`**) các cá thể bào ngư là $(-\infty, 0.52832)$ với độ tin cậy 99%

>> Khoảng tin cậy phải:

In [7]:
print(c(xbar - qt(.99, df = n-1)*SE, Inf))

[1] 0.51967     Inf


**Kết luận:** Khoảng tin cậy đối phải cho giá trị trung bình của chiều dài (cột **`length`**) các cá thể bào ngư là $(0.51967, +\infty)$ với độ tin cậy 99%

>* **Cách 2**: Sử dụng hàm **`t.test()`**

>> Khoảng tin cậy đối xứng:

In [8]:
print(t.test(x = abalone$length, conf.level = 0.99)$conf.int)

[1] 0.51920 0.52878
attr(,"conf.level")
[1] 0.99


**Kết luận:** Khoảng tin cậy đối xứng cho giá trị trung bình của chiều dài (cột **`length`**) các cá thể bào ngư là $(0.51920, 0.52878)$ với độ tin cậy 99%

>> Khoảng tin cậy trái:

In [9]:
print(t.test(x = abalone$length, alternative = "less", conf.level = 0.99)$conf.int)

[1]    -Inf 0.52832
attr(,"conf.level")
[1] 0.99


**Kết luận:** Khoảng tin cậy trái cho giá trị trung bình của chiều dài (cột **`length`**) các cá thể bào ngư là $(-\infty, 0.52832)$ với độ tin cậy 99%

>> Khoảng tin cậy phải:

In [10]:
print(t.test(x = abalone$length, alternative = "greater", conf.level = 0.99)$conf.int)

[1] 0.51967     Inf
attr(,"conf.level")
[1] 0.99


**Kết luận:** Khoảng tin cậy phải cho giá trị trung bình của chiều dài (cột **`length`**) các cá thể bào ngư là $(0.51967, +\infty)$ với độ tin cậy 99%

### Nhận xét:
Cách tự tính bằng công thức và dùng hàm **`t.test()`** đều cho ra cùng một kết quả.

**b. Phương sai đã biết bằng hàm `z.test()`**

- Áp dụng tìm khoảng tin cậy cho cột **`height`** với độ tin cậy $\gamma = 0.95$.
- Giả sử phương sai $\sigma = 0.2$

>> Khoảng tin cậy đối xứng:

In [11]:
print(z.test(x = abalone$height, sigma.x = 0.04, conf.level = 0.95)$conf.int)

[1] 0.13830 0.14073
attr(,"conf.level")
[1] 0.95


**Kết luận:** Khoảng tin cậy đối xứng cho giá trị trung bình của chiều cao (cột **`heigth`**) các cá thể bào ngư là $(0.13830, 0.14073)$ với độ tin cậy 95%

>> Khoảng tin cậy trái:

In [10]:
print(z.test(x = abalone$height, alternative = "less", sigma.x = 0.04, conf.level = 0.95)$conf.int)

[1]      NA 0.14053
attr(,"conf.level")
[1] 0.95


**Kết luận:** Khoảng tin cậy trái cho giá trị trung bình của chiều cao (cột **`heigth`**) các cá thể bào ngư là $(-\infty, 0.14053)$ với độ tin cậy 95%

>> Khoảng tin cậy phải:

In [11]:
print(z.test(x = abalone$height, alternative = "greater", sigma.x = 0.04, conf.level = 0.95)$conf.int)

[1] 0.1385     NA
attr(,"conf.level")
[1] 0.95


**Kết luận:** Khoảng tin cậy phải cho giá trị trung bình của chiều cao (cột **`heigth`**) các cá thể bào ngư là $(0.1385, +\infty)$ với độ tin cậy 95%

**2. Khoảng tin cậy cho tỷ lệ bằng hàm `prop.test()`**

- Áp dụng tìm khoảng tin cậy cho tỷ lệ cá thể có giới tính là **`Male`** với độ tin cậy $\sigma = 0.95$.

**k** là số cá thể có giới tính **`Male`**.

In [15]:
k = sum(abalone$sex == "M")
sprintf("k = %d", k)

[1] "k = 1528"

>> Khoảng tin cậy đối xứng:

In [16]:
print(prop.test(k, n)$conf.int)

[1] 0.35122 0.38066
attr(,"conf.level")
[1] 0.95


**Kết luận:** Khoảng tin cậy đối xứng cho tỷ lệ các cá thể bào ngư mang giới tính đực (cột **`Male`**) là $(0.35122, 0.38066)$ với độ tin cậy 95%

>> Khoảng tin cậy trái:

In [20]:
print(prop.test(k, n, alternative = "less")$conf.int)

[1] 0.00000 0.37827
attr(,"conf.level")
[1] 0.95


**Kết luận:** Khoảng tin cậy trái cho tỷ lệ các cá thể bào ngư mang giới tính đực (cột **`Male`**) là $(0.00000, 0.37827)$ với độ tin cậy 95%

>> Khoảng tin cậy phải:

In [21]:
print(prop.test(k, n, alternative = "greater")$conf.int)

[1] 0.35353 1.00000
attr(,"conf.level")
[1] 0.95


**Kết luận:** Khoảng tin cậy phải cho tỷ lệ các cá thể bào ngư mang giới tính đực (cột **`Male`**) là $(0.35353, 1.00000)$ với độ tin cậy 95%

**3. Khoảng tin cậy cho phương sai**

- Áp dụng tìm khoảng tin cậy cho cột **`length`** với độ tin cậy $\sigma = 0.99$.

In [23]:
vbar = var(abalone$length)
s = sd(abalone$length)
left = s*s*(n-1)/qchisq(.995, df = n-1)
right = s*s*(n-1)/qchisq(.005, df = n-1)
print(left)
print(right)

[1] 0.013641
[1] 0.015269


- Tạo hàm **var.interval** để tìm khoản ước lượng với với 2 tham số **data, conf.level** (Trường hợp kỳ vọng không biết).

In [24]:
var.interval = function(data, alternative = "two.sided", conf.level = 0.95) {
    df = length(data) - 1
    chi.lower = qchisq((1 - conf.level)/2, df)
    chi.upper = qchisq((1 - conf.level)/2, df, lower.tail = FALSE)
    v = var(data)
    
    res = 1
    if(alternative == "two.sided") {
        res = c(df*v/chi.upper, df*v/chi.lower)
    }
    
    if(alternative == "less") {
        res = c(0, df*v/qchisq(1 - conf.level, df))
    }
    
    if(alternative == "greater") {
        res = c(df*v/qchisq(conf.level, df), Inf)
    }
    
    return(res)
}

* Khoảng tin cậy đối xứng của trung bình 

In [25]:
print(var.interval(abalone$length, alternative = "two.sided", conf.level = 0.99))

[1] 0.013641 0.015269


**Kết luận:** Khoảng tin cậy đối xứng cho phương sai của chiều dài (cột **`length`**) các cá thể bào ngư là $(0.013641, 0.015269)$ với độ tin cậy 99%

## II. Tìm cỡ mẫu:

**1. Trường hợp ước lượng cho giá trị trung bình:**

**a. Đã biết phương sai:**
- Viết hàm **sample.size.mean.fi** trả về **cỡ mẫu tối thiểu** để ước lượng được **giá trị trung bình** với sai số không quá **`margin.error`** và độ tin cậy **`conf.level`** (phương sai đã biết là **`sigma`**).
- Hàm nhận vào 3 tham số:
 - **sigma**: phương sai đã biết
 - **margin.error**: sai số
 - **conf.level**: độ tin cậy

In [26]:
sample.size.mean.fi = function(sigma, margin.error, conf.level) {
    return(sigma^2 * qnorm(1 - (1 - conf.level)/2)^2 / margin.error^2)
}

- Ví dụ: phương sai là **9**, sai số là **2**, độ tin cậy là **0.95**

In [27]:
print(sample.size.mean.fi(9, 2, 0.95))

[1] 77.79


**Kết luận:** Cỡ mẫu cần tìm là 78

**b. Chưa biết phương sai**

- Viết hàm **sample.size.mean.se** trả về **cỡ mẫu tối thiểu** để ước lượng được **giá trị trung bình** với sai số không quá **`margin.error`** và độ tin cậy **`conf.level`** (kích thước mẫu sơ bộ là **`num.sample`**).
- Hàm nhận vào 4 tham số:
 - **data**: cột vector cần ước lượng
 - **num.sample**: kích thước mẫu sơ bộ
 - **margin.error**: sai số
 - **conf.level**: độ tin cậy

In [21]:

sample.size.mean.se = function(data, num.sample, margin.error, conf.level) {
    new.sample = sample(data, num.sample) 
    sigma = sd(new.sample)
    res = ceiling(sigma^2 * qnorm(1 - (1 - conf.level)/2)^2 / margin.error^2)
    add.res = res - num.sample
    return(c(res, add.res))
}


Áp dụng với cột **`length`** với cỡ mẫu sơ bộ **100**, sai số **0.01**, độ tin cậy **0.95**
- Tìm cỡ mẫu với 100 mẫu sơ bộ có kích thước 200, ta thu được cỡ mẫu trung bình cần tìm:

In [76]:
get.sample.size = function(times, data, num.sample, margin.error, conf.level) {
    sample.size = c()
    for(it in 1:times) {
        temp = sample.size.mean.se(data, num.sample, margin.error, conf.level)[1]
        sample.size = c(sample.size, temp)
    }
    print("Khoang tin cay cua co mau voi do tin cay 0.99:")
    print(t.test(x = sample.size, alternative = "two.sided", conf.level = 0.99)$conf.int[1])
    print(t.test(x = sample.size, alternative = "two.sided", conf.level = 0.99)$conf.int[2])
    return(mean(sample.size))
}

answer = get.sample.size(100, abalone$length, 200, 0.01, 0.95)
sprintf("Co mau trung binh can tim: %d", ceiling(answer))

[1] "Khoang tin cay cua co mau voi do tin cay 0.99:"
[1] 536.34
[1] 564.62


[1] "Co mau trung binh can tim: 551"

>> - Ta thu được khoảng tin cậy của cỡ mẫu với độ tin cậy 0.99 là **(536.34; 564.62)**

>> - Cỡ mẫu trung bình cần tìm là: **551**

**2. Trường hợp ước lượng cho tỷ lệ:**

- Viết hàm **sample.size.prop** trả về **cỡ mẫu tối thiểu** để ước lượng được **tỷ lệ** với sai số không quá **`margin.error`** và độ tin cậy **`conf.level`** (kích thước mẫu sơ bộ là **`num.sample`**).
- Hàm nhận vào 4 tham số:
 - **attribute**: tính chất A cần xét
 - **data**: cột vector cần ước lượng
 - **num.sample**: kích thước mẫu sơ bộ
 - **margin.error**: sai số
 - **conf.level**: độ tin cậy

In [46]:
sample.size.prop <- function(data, attribute, num.sample, margin.eror, conf.level) {
    p.k = 0
    
    while(TRUE) {
        new.sample = sample(data, num.sample)
        k = sum(new.sample == attribute)
        p.k = k/n

        if(p.k*num.sample >= 5 && (1-p.k)*num.sample >= 5) {
            break
        }
    }
    
    return(qnorm(1 - (1 - conf.level)/2)^2 * p.k * (1 - p.k) / margin.eror^2)
}

In [47]:
print(sample.size.prop(abalone$sex, "M", 400, 0.01, 0.95))

[1] 1287.2


In [70]:
get.sample.size.prop = function(times, data, attribute, num.sample, margin.error, conf.level) {
    sample.size = c()
    for(it in 1:times) {
        temp = sample.size.prop(data, attribute, num.sample, margin.error, conf.level)
        sample.size = c(sample.size, temp)
    }
    # print(sample.size)
    print("Khoang tin cay cua co mau voi do tin cay 0.99:")
    print(t.test(x = sample.size, alternative = "two.sided", conf.level = 0.99)$conf.int[1])
    print(t.test(x = sample.size, alternative = "two.sided", conf.level = 0.99)$conf.int[2])
    return(mean(sample.size))
}


Tìm cỡ mẫu với 200 mẫu sơ bộ có kích thước 400, ta thu được cỡ mẫu trung bình cần tìm:

In [80]:
answer = get.sample.size.prop(200, abalone$sex, "M", 400, 0.01, 0.95)
sprintf("Co mau trung binh can tim: %d", ceiling(answer))

[1] "Khoang tin cay cua co mau voi do tin cay 0.99:"
[1] 1286.3
[1] 1316.3


[1] "Co mau trung binh can tim: 1302"

>> - Ta thu được khoảng tin cậy của cỡ mẫu với độ tin cậy 0.99 là **(1286.3; 1316.3)**

>> - Cỡ mẫu trung bình cần tìm là: **1302**

- Áp dụng công thức: $n \geq \dfrac{(u_{1-\alpha/2})^2}{4\epsilon^2}$
 - Với sai số **0.01** và độ tin cậy **0.95**, ta thu được cỡ mẫu:

In [79]:
margin.error = 0.01
conf.level = 0.95
answer = qnorm(1 - (1 - conf.level)/2)^2 / (4*margin.error^2)
print(answer)

[1] 9603.6


>> Ta thu được cỡ mẫu tối thiểu cần thiết là **9604**

### Nhận xét:
Trong trường hợp của bài toàn này, cỡ mẫu tìm được bằng hàm tự việt tối ưu hơn so với dùng công thức Cauchy (1302 so với 9604).